services/MerchStoreService.ts

```diff
@@ -291,6 +291,68 @@ export default class MerchStoreService {
       await user.reload();
       const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
       const itemOptions = await merchItemOptionRepository.batchFindByUuid(originalOrder.map((oi) => oi.option));
+
+      const totalCost = await this.verifyOrder(originalOrder, user);
+
+      const merchOrderRepository = Repositories.merchOrder(txn);
+
+      // if all checks pass, the order is placed
+      const createdOrder = await merchOrderRepository.createMerchOrder(OrderModel.create({
+        user,
+        totalCost,
+        items: flatten(originalOrder.map((optionAndQuantity) => {
+          const option = itemOptions.get(optionAndQuantity.option);
+          const quantityRequested = optionAndQuantity.quantity;
+          return Array(quantityRequested).fill(OrderItemModel.create({
+            option,
+            salePriceAtPurchase: option.getPrice(),
+            discountPercentageAtPurchase: option.discountPercentage,
+          }));
+        })),
+      }));
+
+      const activityRepository = Repositories.activity(txn);
+      await activityRepository.logActivity({
+        user,
+        type: ActivityType.ORDER_MERCHANDISE,
+        description: `Order ${createdOrder.uuid}`,
+      });
+
+      await Promise.all(originalOrder.map(async (optionAndQuantity) => {
+        const option = itemOptions.get(optionAndQuantity.option);
+        const updatedQuantity = option.quantity - optionAndQuantity.quantity;
+        return merchItemOptionRepository.upsertMerchItemOption(option, { quantity: updatedQuantity });
+      }));
+
+      const userRepository = Repositories.user(txn);
+      userRepository.upsertUser(user, { credits: user.credits - totalCost });
+
+      return [createdOrder, itemOptions];
```

Comment:
I know I wrote this part but this is terrible. We should probably have some proper return type for this.

Reply:
Is there a good description for the return type here?

Reply:
Maybe `MerchOrderAndItemOptions`? You can add this type to `types/internal/index.ts`.